poly/promise/promise.go

```diff
@@ -0,0 +1,744 @@
+	//   2. The length of the signature
+	//   3. The signature itself
+	// 1 and 2 are variable length but no more than an int64. Hence, enough
+	// can be read to unmarshal 1 and 2. After that is done, the length of
+	// the entire object can be recomputed. Hence, the proper buffer can
+	// be made and the rest of the unmarshalling can be handled by
+	// BinaryUnmarshal.
+	
+	// Retrieve the signature index as stated above.
+	intSize    :=  binary.Size(int64(p.pi))
+	buf := make([]byte, intSize*2)
+	n, err := io.ReadFull(r, buf)
+
+	if err != nil {
+		return n, err
+	}
+	
+	// Find the length of the signature index as stated above.
+	_, piLen := binary.Varint(buf)
+	if piLen <= 0 {
+		return 0, errors.New("Binary number not provided for index")
+	}
+	
+	// Retrieve the signature length
+	sigLen, sigLenBytesRead := binary.Varint(buf[piLen:])
+	if sigLenBytesRead <= 0 {
+		return 0, errors.New("Signature length not provided")
+	}
+
+	// The entire length of the buffer can now be calculated and the
+	// object = unmarshalled.
+	finalBuf := make([]byte, int(sigLen) + piLen + sigLenBytesRead)
+	
+	// Copy what has already been read into the buffer.
+	copy(finalBuf, buf)
+	
+	// Read the rest.
+	m, err2 := io.ReadFull(r, finalBuf[n:])
+	if err2 != nil {
+		return m, err2
+	}
+	return n+m, p.UnmarshalBinary(finalBuf)
+}
+
+/* PromiseShare is used to represent a secret share of a Promise. When a
+ * guardian wants to reveal the secret share it is guarding either to a promised
+ * client or to other guardians to prove that the promiser is crooked, it
+ * constructs this object.
+ *
+ * The key features of a PromiseShare are:
+ *	1. The index of the share
+ *      2. The share itself
+ *      3. The Diffie-Hellman secret between the guardian and promiser
+ *
+ * #3 is used for verification purposes. Other servers can use it to prove that
+ * encrypting #2 with #3 will produce the secret stored in the promise.
+ *
+ * As mentioned above, PromiseShare's can also be used as "BlameProof" objects.
+ *
+ * The BlameProof object provides an accountability measure. If a promiser
+ * decides to construct a faulty share, guardians can construct a BlameProof
+ * to show that the server is malicious. 
+ * 
+ * The guardian provides the index of the bad secret, the bad secret itself,
+ * and its diffie-hellman shared key with the promiser. Other servers can then
+ * verify if the promiser is malicious or the guardian is falsely accusing the
+ * server.
+ *
+ * To quickly summarize the blame procedure, two things must hold for the blame
+ * to succeed:
+ *
+ *   1. The provided share when encrypted with the diffie key must equal the
+ *   point provided at index bi of the promise. This verifies that the share
+ *   was actually intended for the guardian.
+ *
+ *   2. The provided share must fail to pass pubPoly.Check. This ensures that
+ *   the share is actually corrupted and the guardian is not just lying.
+ */
+
+type PromiseShare struct {
```

Comment:
It's not yet clear to me exactly what role(s) this PromiseShare type is supposed to serve.  It doesn't like the Promise struct directly contains or refers to this type.  What exactly is it for?